.github/ci-scripts/job_runner.py

```diff
@@ -0,0 +1,123 @@
+import argparse
+import asyncio
+import json
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import Optional
+
+import duckdb
+from ray.job_submission import JobStatus, JobSubmissionClient
+
+
+def parse_env_var_str(env_var_str: str) -> dict:
+    iter = map(
+        lambda s: s.strip().split("="),
+        filter(lambda s: s, env_var_str.split(",")),
+    )
+    return {k: v for k, v in iter}
+
+
+async def print_logs(logs):
+    async for lines in logs:
+        print(lines, end="")
+
+
+async def wait_on_job(logs, timeout_s):
+    await asyncio.wait_for(print_logs(logs), timeout=timeout_s)
+
+
+def generate_data():
+    datadir = Path(__file__).parents[2] / "gendata"
+    datadir.mkdir(parents=True, exist_ok=True)
+    scale_factor = 0.01
+    db = duckdb.connect(database=datadir / "tpcds.db")
+    db.sql(f"call dsdgen(sf = {scale_factor})")
+    for item in db.sql("show tables").fetchall():
+        tbl = item[0]
+        parquet_file = datadir / f"{tbl}.parquet"
+        print(f"Exporting {tbl} to {parquet_file}")
+        db.sql(f"COPY {tbl} TO '{parquet_file}'")
```

Comment:
Is this not TPC-DS specific functionality? Why is this inside the generic `job_runner.py`. This means that every time I run my non-TPC-DS related jobs I will trigger data generation?

Doing so also adds a `duckdb` dependency to all users of your workflow for no reason.

Reply:
If we want one singular entrypoint script that is running on the GH runner (as opposed to a generic script that the end-user can supply to us), we'll have to "massage" it into fitting as many different use-cases for us.

One way to avoid data-generation would be to supply command line arguments to this script. For that, we could introduce a new GHA input-parameter, named `runner_args` or something, which is specifically passed to this GH runner script. We can then parse and execute conditional logic based off of the contents of that `runner_args` parameter.

Reply:
That's really not true, and there are many possible solutions around that. As it stands, the data generated here shouldn't even be accessible from the cluster -- I'm not really sure how the cluster would execute a query on this data.

1. Enforce that data generation must happen outside of this workflow (people should generate their data and put it into S3 so that it's accessible from the cluster)
2. Add an optional pre-job script argument that users can run

I would opt for (1).

Reply:
The way the data is being accessed by the cluster is that it's being generated locally and then being submitted up. Sure, this is not ideal since that will be a huge amount of data for more serious workloads.

The way to implement this would be by using `aws s3 sync` to modify the templatization of the `.github/assets/template.yaml` file to run `aws s3 sync ...` as one of its setup commands. Then, the data is accessible only to the cluster and not generated locally.

That will require the templatization script to be updated (for example, with a flag `--link-awss3-tpcds-data`) that adds a `aws s3 sync ...` command if the flag exists.

Reply:
That will require the tempatization step to be updated with a new flag. This flag will then need to be exposed somehow to the `run-cluser` workflow.

Reply:
Resolved offline.

Reply:
No data generation being performed, everything is directly read in from S3.